src/ShelfLayoutManager.Infrastructure/CabinetsDataService.cs

```diff
@@ -0,0 +1,116 @@
+using Microsoft.EntityFrameworkCore;
+using ShelfLayoutManager.Core;
+using ShelfLayoutManager.Infrastructure.Converters;
+
+namespace ShelfLayoutManager.Infrastructure;
+
+/// <inheritdoc>
+public class CabinetsDataService(DbContext dbContext, ICabinetEntityConverter cabinetEntityConverter)
+    : ICabinetsDataService
+{
+    private readonly DbContext _dbContext = dbContext;
+
+    private readonly ICabinetEntityConverter _cabinetEntityConverter = cabinetEntityConverter;
+
+    public async Task<Result<Cabinet?>> GetAsync(long id, bool includeRows, bool includeLanes, CancellationToken ct) =>
+        _cabinetEntityConverter.ConvertToCabinet(
+            await GetReadQueryable(includeRows, includeLanes).Where(c => c.Id == id).FirstOrDefaultAsync(ct));
+
+    public async Task<Result<IList<Cabinet>>> ListAsync(bool includeRows, bool includeLanes, CancellationToken ct) =>
+        (await GetReadQueryable(includeRows, includeLanes).ToListAsync(ct))
+            .Select(_cabinetEntityConverter.ConvertToCabinet).OfType<Cabinet>().ToArray();
+
+    public async Task<Result> AssignSkuAsync(CabinetSkuAssignment cabinetSkuAssignment, CancellationToken ct)
+    {
+        // TODO(abalkar): Add validation for given assignment.
+
+        bool getFromLane = cabinetSkuAssignment.FromRowNumber != 0 || cabinetSkuAssignment.FromLaneNumber != 0;
+        bool getToLane = cabinetSkuAssignment.ToRowNumber != 0 || cabinetSkuAssignment.ToLaneNumber != 0;
+
+        LaneEntity? fromLaneEntity = null;
+        LaneEntity? toLaneEntity = null;
+
+        if (getFromLane)
+        {
+            fromLaneEntity = await GetLane(
+                cabinetSkuAssignment.CabinetNumber,
+                cabinetSkuAssignment.FromRowNumber,
+                cabinetSkuAssignment.FromLaneNumber,
+                ct);
+
+            if (fromLaneEntity is null)
+            {
+                return Result.Fail("Could not find from lane with given numbers.");
+            }
+        }
+
+        if (getToLane)
+        {
+            toLaneEntity = await GetLane(
+                cabinetSkuAssignment.CabinetNumber,
+                cabinetSkuAssignment.ToRowNumber,
+                cabinetSkuAssignment.ToLaneNumber,
+                ct);
+
+            if (toLaneEntity is null)
+            {
+                return Result.Fail("Could not find to lane with given numbers.");
+            }
+        }
+
+        bool removingSku = getFromLane && !getToLane;
+        bool assigningSku = !getFromLane && getToLane;
+        bool movingSku = getFromLane && getToLane;
+
+        if (removingSku)
+        {
+            fromLaneEntity!.JanCode = null;
+        }
+        else if (assigningSku)
+        {
+            toLaneEntity!.JanCode = cabinetSkuAssignment.JanCode;
+        }
+        else if (movingSku)
+        {
+            toLaneEntity!.JanCode = fromLaneEntity!.JanCode;
+            toLaneEntity!.Quantity = fromLaneEntity!.Quantity;
+
+            fromLaneEntity.JanCode = null;
```

Comment:
Nit: Could set Quantity to zero, but I suppose Quantity has no meaning when JanCode is null